feat(gnovm): Executable Markdown

notJoon opened the pull request:
# Description

The `doctest` (tentative name) is a tool that allows code blocks in Markdown, especially those marked as "go" and "gno", to be executed using the gno VM.

It works as follows: First, it parses the code blocks (fenced code blocks) from the Markdown file. After that, the extracted code is written to a temporary file, and the code is executed at the VM level.

Also supports static analysis to automatically include imports or packages without explicitly specifying them before execution. However, in this case, it assumes the code always executes the main function, fixes the package as 'main', and only searches for missing library paths in the standard library.
Additionally, to prevent unnecessary executions, a feature to cache execution results has been added.

I wrote the behavior and functionality in the file `gnovm/cmd/gno/testdata/doctest/1.md` as a tutorial.

Moving forward, there is potential to further enhance this tool to support code documentation by allowing the execution of code embedded within comments, similar to Rust.

related #2245 

## CLI

The CLI command is used as follows:

> gno doctest -path <markdown_file_path> [-run <pattern>] [-timeout <duration>]

`path` literally means the path of the markdown file to be executed. The `run` flag allows you to specify a pattern to match against code block names or tags, similar to `go test -run <name>`. This will execute all code blocks that partially or completely match the given pattern. If `run` is not specified, all code blocks in the file will be executed.

The optional `timeout` flag allows you to set a maximum duration for the execution of each code block. If not specified, a default timeout will be used.

For example:
> gno doctest -path ./README.md -run "example" -timeout 30s

This command will execute all code blocks in README.md that have "example" in their name or tag, with a timeout of 30 seconds for each block.

moul commented on gnovm/cmd/gno/testdata/doctest/1.md:
```diff
@@ -0,0 +1,87 @@
+import "strings"
+
+func main() {
+    println(strings.ToUpper("Hello, World"))
```
i don't get where you're putting the output?

notJoon replied:
The execution results are not separately saved (although they are stored in the cache) and are only shown through the CLI. At first, I thought this would be sufficient, but it seems that saving the values would be more useful.

moul commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,179 @@
+// ExecuteCodeBlock executes a parsed code block and executes it in a gno VM.
+func ExecuteCodeBlock(c codeBlock, stdlibDir string) (string, error) {
+	if c.lang == "go" {
+		c.lang = "gno"
```
since we don't have gno fully recognized by github atm, i suggest we start having something like

````
```go,gnodoctest
package foo

func main() {
	a := 42
}
```
````

notJoon replied:
It sounds good. I think we could consider setting execution options like `ignore` or `should_panic` for that feature.

notJoon replied:
Changing 'go' to 'gno' (or using `gno`) seems fine as it's intended to specify the file extension for temporary executable files. 

However, I'm slightly concerned that including multiple elements on a single line when specifying the tag might break syntax highlighting during document rendering, especially using `go` language tag. 

If this isn't an issue, it might be useful to support specifying execution options like `ignore` or `should_panic`.

deelawn reviewed (commented):
Nice work -- I found this very well written and and easy to follow, but there is a lot here, so I've left plenty of comments regarding style and questions about certain things. Thanks for this 😄

notJoon commented on gnovm/cmd/gno/doctest_test.go:
```diff
@@ -0,0 +1,98 @@
+)
+
+func TestDoctest(t *testing.T) {
+	t.Skip("skipping test")
```
Oh I forgot to remove that line 😅

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+	for _, imp := range f.Imports {
+		if imp.Name != nil {
+			imports = append(imports, imp.Name.Name)
+		} else {
```
I should add this pattern as a lint check rule later

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+	if prefix != "" {
+		parts = append(parts, prefix)
+	}
+	if len(imports) > 0 {
```
Moving the imports to the end is better than previous way. If I think of a more better way to generate names, I'll nned to improve this part later.

notJoon replied:
That's the good point. actually I didn't thought about that. If I think of a way to generate better names, I should improve this part further.

notJoon commented on gnovm/cmd/gno/doctest.go:
```diff
@@ -0,0 +1,109 @@
+		"",
+		"pattern to match code block names",
+	)
+	fs.Duration(
```
updated https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

deelawn replied:
I think you still need to assign it to `c.timeout` though.

deelawn replied:
Never mind, I see this was changed in the next commit.

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+		switch strings.TrimSpace(option) {
+		case "ignore":
+			options.Ignore = true
+		case "should_panic":
```
https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+
+	// parser options from the first line of the code block
+	if bytes.HasPrefix(firstLine, []byte("//")) {
+		re := regexp.MustCompile(`@(\w+)(?:="([^"]*)")?`)
```
Added comments about what is parsed and processed.

https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+
+	prefix := determinePrefix(f)
+	imports := extractImports(f)
+	expectedOutput, _, _ := parseExpectedResults(content)
```
https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+// The name is constructed based on the code's prefix (Test, Print or Calc),
+// imported packages, main identifier, and expected output.
+func generateCodeBlockName(content string) string {
+	lines := strings.Split(content, "\n")
```
fixed both `generateCodeBlockName` and `generateFallbackName` functions to use `bufio`. thanks!

https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+// containsPrintStmt checks if the given function declaration contains
+// any print or println statements.
+func containsPrintStmt(fn *ast.FuncDecl) bool {
+	var yes bool
```
https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+// containsCalculation checks if the given function declaration contains
+// any binary or unary expressions, which are indicative of calculations.
+func containsCalculation(fn *ast.FuncDecl) bool {
+	var yes bool
```
change ok variable name to `hasCalcStmt`

https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+	// Extract the actual language from the lang field
+	lang := strings.Split(c.lang, ",")[0]
+
+	if lang != "go" && lang != "gno" {
```
https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+	acc := acck.NewAccountWithAddress(ctx, addr)
+	acck.SetAccount(ctx, acc)
+
+	coins := std.MustParseCoins("")
```
https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+
+type ExecutionOption struct {
+	Ignore      bool
+	ShouldPanic string
```
I intended to use the execution option names directly as field names, but it was a bit ambiguous.

https://github.com/gnolang/gno/pull/2357/commits/6bac3adcea48204cfb2dff4adcaad3d870afa521

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+		language = "plain"
+	}
+
+	firstLine := body[lines.At(0).Start:lines.At(0).Stop]
```
It's possible when you don't specify a language for the code block (for example, using only ``` at the start when declaring a code block).

It's not a major issue, but I added this line to clearly categorize the language.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	compiledPattern := regexp.QuoteMeta(pattern)
+	compiledPattern = strings.ReplaceAll(compiledPattern, "\\*", ".*")
```
I added these lines to convert the original pattern string into a valid form for regular expressions.

For example, if a pattern like `foo*.md` comes in, it converts it to `foo.*\.md`.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				result, err := ExecuteCodeBlock(block, GetStdlibsDir())
```
Yes, that's correct. I wrote it this way to improve speed in cases where there are multiple code blocks matching the pattern, or when executing the entire code block.

deelawn replied:
I don't understand.  Why does there need to be locking if `ExecuteCodeBlock` is only ever executed in a single go routine?

notJoon replied:
Checking again, it was unnecesary. So, I remove the locks

1. https://github.com/gnolang/gno/pull/2357/commits/2d64669b6bc052cbb8e16ca85c4c36a0cafc899d (remove lock)

2. https://github.com/gnolang/gno/pull/2357/commits/317fd41c20cf0837eb618d382cd64bd1a80bf7ce (remove lock from cache)

notJoon commented on gnovm/pkg/doctest/analyzer.go:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	ensurePackageDeclaration(node)
+	if exist := ensureMainFunction(node); !exist {
```
I removed the `ensurePackageDeclaration` function as it definitely not necessary as a separate function, but I left `ensureMainFunction` as is to keep the function size small.

However, as you mentioned, I changed the function name to `hasMainFunction`.

notJoon commented on gnovm/pkg/doctest/parser.go:
```diff
@@ -0,0 +1,350 @@
+
+	var outputs, errors []string
+
+	cleanSection := func(section string) string {
```
Not really. Still I cannot think any other reasons, so it does not a big deal. removed

https://github.com/gnolang/gno/pull/2357/commits/4978329b768d8849c8daad0d201e09c40ea10dd3

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+
+// GetStdlibsDir returns the path to the standard libraries directory.
+func GetStdlibsDir() string {
+	_, filename, _, ok := runtime.Caller(0)
```
I made it search for the location in the gnovm directory like that because other methods couldn't recognize the stdlib.

When using other methods, I couldn't use such packages. There might be better ways to handle this, but I can't think of any for now.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,246 @@
+		expected = strings.TrimSpace(expectedError)
+	}
+
+	if expected == "" {
```
Initially, to handle simple execution, I decided to output the results even when the expected output was empty. However this seems not clear and inappropriate in a testing context.

So, I've adjusted the funtion to distinguish between simple execution and test purpose by checking the values stored in the `codeBlock`'s `expectedOutput` and `expectedError` fields.

For example, in cases where no expected output is specified, just output the value as before:

```plain
'''go
func main() {
    println("This is an unexpected output")
}
'''
```

However, in cases where an expected output is specified, this must fail if the actual output is differs as expected.

```plain
'''go
func main() {
    println("This is an unexpected output")
}

// Output:
'''
```

https://github.com/gnolang/gno/pull/2357/commits/15770341009f96815b3b0dea5bda512864d53c1d

notJoon commented on gnovm/pkg/doctest/analyzer.go:
```diff
@@ -0,0 +1,205 @@
+			return true
+		}
+
+		if remainingPackages[ident.Name] {
```
Very tricky one. I think the approach could be to fist traverse the AST to collect local identifiers (functions, variables...), and then traverse AST again to examine each node, treating an identifier as the stdlib only if it's not in the list of local identifiers. 🤔🤔🤔🤔

notJoon replied:
it works https://github.com/gnolang/gno/pull/2357/commits/317fd41c20cf0837eb618d382cd64bd1a80bf7ce

deelawn replied:
I think that, in order to know whether or not to add a stdlib import, we need to know the local identifiers that are defined in the current scope. The current scope would be any package level identifiers, block level identifiers, and ancestor block level identifiers (I think?).  For example, this code will import the `math` stdlib when it should not:
```go
package main
			
type mmm struct{
	MaxUint64 uint64
}

func main() {
	maxValue := getMaxValue()
	println(maxValue)
}

func getMaxValue() uint64 {
	math := mmm{MaxUint64: 1<<64 - 1}
	return math.MaxUint64
}
```

deelawn replied:
I wonder if these feature should be omitted for the time being, making it required that people include the necessary imports when writing code. This is a nice to have but I think it could be difficult to do correctly.

notJoon replied:
It would be nice if it could be handled well, but there certainly seems to be some risk. So, as mentioned below, I think it's appropriate to exclude this part. So will removed.

notJoon commented:
Thank you for the detailed review @deelawn. I think I have addressed all the points you mentioned in your review. Could you please check and confirm? Thank you once again 👍

piux2 commented on gnovm/pkg/doctest/README.md:
````diff
@@ -0,0 +1,131 @@
+
+Doctest also recognizes that a block of code is a gno. The code below outputs the same result as the example above.
+
+```go
````
How does Doctest detect this is a gno? because //@test: print hello world?

What is the reason behind using "```go" instead of  "````gno"  for gno code block?

notJoon replied:
That's a really good question. Currrently, when rendering markdown, gno doesn't have universal syntax highlighting applied yet (except the custom syntax rule). So, even in official documents like "Effective gno" are using `go` instead of `gno` to specify the language.

Considering the purpose of markdown is more for documentation rather than testing, this was an inevitable purpose. So I made it recognize both languages.

However, it doesn't clearly distinguish whether it's completely `go` or `gno`. For now, I assume that it's mostly compatible with go and execute it at gno VM level. In other words, even if the language is atually written in go, everything runs as gno. This part may require deeper context analysis in the future.

To summarize in the order of your questions:
  1. Even if the code block is specified as Go, it's actually recognized and executed as gno.
  2. There's no feature to clearly distinguish between the two languages, and it relies on the language specified in the fenced code block.
  3. Go is also set to be recognized due to syntax highlighting.

piux2 replied:
The main difference between Gno and Go is that Gno has a few native methods unique to blockchain usage. For example:

[std.AssertOriginCall](https://github.com/gnolang/gno/blob/912a5dbf5c1d7118472a4f46b26bfcd7f4072856/gnovm/stdlibs/generated.go#L422)
[std.GetHeight](https://github.com/gnolang/gno/blob/912a5dbf5c1d7118472a4f46b26bfcd7f4072856/gnovm/stdlibs/generated.go#L474C4-L474C13)

Basically, a program can either be Go code or Gno code. We cannot mix them in one source file.

There are a few options I can think of

- Users need to explicitly mark it as either Go code or Gno code, similar to indicating the source code type with a suffix in the filename: xxx.go vs xxx.gno.
- Only support Gno.
- By default, the MD is treated as Gno code. Use gnostd.XXX to indicate that it is Go code. (Not sure if this has much practical purpose; it is more complicated to implement and is only for the sake of supporting Go code with chain functionality in MD.)

piux2 commented on gnovm/pkg/doctest/README.md:
```diff
@@ -0,0 +1,131 @@
+package main
+
+func main() {
+    println("Hello, World!")
```
the println does not print correct results compare with file tests.


  package main
  type ints []int
  func main() {
        a := ints{1,2,3}
      println(a)
  }

doctest

// Output:
// (slice[(1 int),(2 int),(3 int)] gno.land/r/g14ch5q26mhx3jk5cxl88t278nper264ces4m8nt/run.ints)

file test result is correct ints is a type defined in main package. 
 // Output:
 // (slice[(1 int),(2 int),(3 int)] main.ints)

notJoon replied:
I think it print out the original caller's address instead of the main. It seems to be related to the implementation of `ExecutionCodeBlock` in `exec.go`. I'll need to look at it. thanks!

piux2 replied:
Maybe we can replace it with the proper type when displaying it in the Markdown results.

piux2 commented on gnovm/pkg/doctest/README.md:
```diff
@@ -0,0 +1,131 @@
+
+## 4. Omitting Package Declaration
+
+Doctest can even handle cases where the `package` declaration is omitted.
```
Not sure about this, isn't importing package explicitly before their use part of the go and gno specs?

notJoon replied:
I added this feature to allow execution of code snippet that does not have package declaration or import statements. It was somewhat added for conenience, but it does not fit with the purpose of testing. I'll remove this part after some considerations.

deelawn reviewed (approved):
Thanks for the work to resolve my comments.  Everything looks good to me.

zivkovicmilos reviewed (approved):
Seems good, at least from my domain knowledge of the VM 💯 

Please wait on @thehowl to give you a ✅ before moving forward 🙏 

cc @mvertes for a look also

thehowl commented on gnovm/pkg/gnolang/nodes_test.go:
```diff
@@ -1,10 +1,13 @@
-package gnolang_test
```
Nah

thehowl commented:
@notJoon ping me once you've addressed the comments from milos, and i'll take a look :)

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: notJoon/gno-core)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

notJoon commented:
@thehowl 
it's been slightly delayed due to company work recently, I'll ping you soon. 👍

thehowl commented:
ping again for @notJoon

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,319 @@
+}
+
+// ref: gno.land/pkg/sdk/vm/common_test.go
+func setupEnv() (
```
The `setupEnv` function is used not only for testing but also for creating execution environments within `ExecuteCodeBlock`, particularly for recognizing the standard library.

notJoon commented on gnovm/pkg/doctest/parser_test.go:
```diff
@@ -0,0 +1,434 @@
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
Yep, but this line was added because golangci-lint caught it as a lint error. I'll check if the linter's specification has changed and remove it.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,319 @@
+		return "", err
+	}
+
+	cache.set(hashKey, res)
```
There aren't many cases where someone would need to execute the same code block multiple times to make heavy use of caching.

However, when implementing this feature, I referenced the caching capabilities in Rust's doctest and Go's test functionality, and thought it would be useful to include it.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,319 @@
+
+// hashCodeBlock generates a SHA256 hash for the given code block.
+func hashCodeBlock(c codeBlock) string {
+	h := sha256.New()
```
Definitely overkill. If we keep the caching feature, I should make it lighter.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,319 @@
+	cache      = newCache(maxCacheSize)
+	regexCache = make(map[string]*regexp.Regexp)
+
+	addrRegex = regexp.MustCompile(`gno\.land/r/g[a-z0-9]+/[a-z.]+`)
```
Yes, similar regex defined in `gnovm/pkg/gnolang/helpers.go` file, but I couldn't use it directly because it's slightly different.

notJoon commented on gnovm/cmd/gno/doctest.go:
```diff
@@ -0,0 +1,107 @@
+	resultChan := make(chan []string)
+	errChan := make(chan error)
+
+	go func() {
```
fixed https://github.com/gnolang/gno/pull/2357/commits/691d60d4064cc887b2f43d1cbfa113b553becd8f

notJoon replied:
https://github.com/gnolang/gno/pull/2357/commits/691d60d4064cc887b2f43d1cbfa113b553becd8f

notJoon commented on gnovm/cmd/gno/doctest.go:
```diff
@@ -0,0 +1,107 @@
+
+func execDoctest(cfg *doctestCfg, _ []string, io commands.IO) error {
+	if cfg.markdownPath == "" {
+		return fmt.Errorf("markdown file path is required")
```
This part was removed in this commit https://github.com/gnolang/gno/pull/2357/commits/691d60d4064cc887b2f43d1cbfa113b553becd8f

notJoon replied:
https://github.com/gnolang/gno/pull/2357/commits/148300d035a8c901acb2648062afb2080ec6c456

notJoon commented on gnovm/cmd/gno/doctest.go:
```diff
@@ -0,0 +1,107 @@
+	case err := <-errChan:
+		return fmt.Errorf("failed to execute code block: %w", err)
+	case <-ctx.Done():
+		return fmt.Errorf("execution timed out after %v", cfg.timeout)
```
this code block also has been removed. https://github.com/gnolang/gno/commit/691d60d4064cc887b2f43d1cbfa113b553becd8f

notJoon commented:
@zivkovicmilos @thehowl 
I finally fixed this after leaving it covered in the dust for a long time. If you have any additional suggestions, please check and let me know. 

Additionally, CI is currently failing due to go generate issues. I haven't committed the code because generating code in the softfloat directory produces slightly different code from what's in the master branch.

notJoon commented on gnovm/pkg/doctest/exec.go:
```diff
@@ -0,0 +1,319 @@
+		{Name: fmt.Sprintf("%d.%s", c.index, lang), Body: c.content},
+	}
+
+	addr := crypto.AddressFromPreimage([]byte("addr1"))
```
modified to use `GenPrivKey` to generate new addresses.
https://github.com/gnolang/gno/pull/2357/commits/691d60d4064cc887b2f43d1cbfa113b553becd8f#diff-09e28bec44139f0f8166ce0989313dfc3a7f9b4a8ece616c3625e6216cb3f929

notJoon committed:
feat(gnovm): add doctest markdown code block parser

notJoon committed:
feat(gnovm): add doctest executor for parsed code blocks

notJoon committed:
feat(gnovm): add gno doctest subcommand

notJoon committed:
fix ci failures

notJoon committed:
refactor(doctest): replace sdk/vm keeper setup with gnovm/pkg/test

notJoon committed:
refactor(doctest): drop AST-based code block name generator

notJoon committed:
chore(doctest): remove dead codes

notJoon committed:
fix(doctest): share test.ProdStore across blocks

notJoon committed:
refactor(doctest): align directive syntax with filetest

jeronimoalbi commented on gnovm/pkg/doctest/README.md:
```diff
@@ -0,0 +1,97 @@
+includes `gnodoctest` (e.g. `go,gnodoctest`, useful when you want
+GitHub to render Go syntax highlighting for the snippet).
+
+## Directives
+
+Doctest reuses [filetest](../test/filetest.go)'s directive grammar.
+PascalCase keys (`Output`, `Error`) start a multi-line section
+captured from the following `// ...` comment lines until the next
+section, a bare `//`, or any non-comment line. ALLCAPS keys are
+single-line: `KEY:` is a flag, `KEY: value` carries a value.
+
+| Directive | Form | Purpose |
+| --------- | ---- | ------- |
+| `Output` | multi-line | Expected stdout. |
+| `Error`  | multi-line | Expected error/panic text. |
+| `NAME`   | `NAME: <name>` | Block name used by `-run`. Defaults to `block_<index>`. |
+| `IGNORE` | `IGNORE:` | Skip execution of this block. |
+| `SHOULD_PANIC` | `SHOULD_PANIC:` or `SHOULD_PANIC: <msg>` | Block must panic. With a value, the panic must contain that substring. |
```
It might be interesting to have some context on the comments, about why they are there, for people reading the snippets that are not aware they are intended for `gno doctest`.

For example instead of `// IGNORE` something like:

````go
```gno
// gnodoctest: ignore
package main

func main() {}
```
````

Just a though.

jeronimoalbi replied:
It seems `SHOULD_PANIC` might be redundant, given that `Error` also exists, or is there another reason for `SHOULD_PANIC`?

notJoon committed:
Merge branch 'master' into testable-markdown
